plutus-core/plutus-core/src/PlutusCore/Evaluation/Machine/ExMemory.hs

```diff
@@ -164,40 +165,41 @@ instance ExMemoryUsage Char where
     memoryUsage = sizeList
         where sizeList =
                   \case
                    []   -> 0
                    x:xs -> memoryUsage x + sizeList xs
 
-{- Another naive traversal for size.  This only accounts for the number of nodes
- in the Data object, not the sizes of their contents.  I think that's the right
- thing to do, but it means that when we're benchmarking the Data operations
- (especially equalsData) we should do it on objects with simple contents like
- small integers to avoid also measuring costs for eg large bytestrings.  At
- execution time this should be OK because the apppropriate costing functions (eg
- for comparing bytestrings) *will* be run then, and we just have to account for
- the overhead of processing the tree.  It's also important here that arguments
- to costing functions are lazily evaluated, because we don't want traverse a
- Data object when we're doing something like unBData which just looks at the top
- constructor.
+{- Another naive traversal for size.  This accounts for the number of nodes in a
+   Data object, and also the sizes of the contents of the nodes.  This is not
+   ideal, but it seems to be the best we can do.  At present this only comes
+   into play for 'equalsData', which is implemented using the derived
+   implementation of '==' (fortunately the costing functions are lazy, so this
+   won't be called for things like 'unBData' which have constant costing
+   functions because they only have to look at the top node).  The problem is
+   that when we call 'equalsData' the comparison will take place entirely in Haskell,
+   so the costing functions for the contents of 'I' and 'B' nodes won't be called.
+   Thus if we just counted the number of nodes the sizes of 'I 2' and
+   'B <huge bytestring>' would be the same but they'd take different amounts of
+   time to compare.  It's not clear how to trade off the costs of processing a
+   units per node, but we may wish to revise this after experimentationnode and
+   processing the contents of nodes: the implementation below compromises by charging
+   four units per node, but we may wish to revise this after experimentation.
 -}
-
 instance ExMemoryUsage Data where
     memoryUsage = sizeData
         where sizeData =
                 \case
-                 Constr _ l -> 1 + sizeDataList l
-                 Map l      -> 1 + sizeDataPairs l
-                 List l     -> 1 + sizeDataList l
-                 I _        -> 1
-                 B _        -> 1
+                 Constr _ l -> nodeMem + sizeDataList l
+                 Map l      -> nodeMem + sizeDataPairs l
+                 List l     -> nodeMem + sizeDataList l
+                 I n        -> memoryUsage n
+                 B b        -> memoryUsage b
+              nodeMem = 4
               sizeDataList []     = 0
               sizeDataList (d:ds) = sizeData d + sizeDataList ds
               sizeDataPairs []           = 0
```

Comment:
should these zeros be `nodeMem`?

Reply:
Similar question about the instance for list

Reply:
And maybe for the cons constructor too?

Reply:
I think not.  I'm tying to account for the cost of traversing the tree here, and the lists aren't part of the node.  I could have used a fold here but I was in a hurry and I thought it might be quicker (in terms of execution time) just to do the naive thing.  I have however added `nodeMem` to the costs for `I` and `B`.